### PR TITLE
Look for the latest build with a non-matching commit hash

### PR DESCRIPTION
**Pull request by rdner**
Upgrade tests were failing because of the matching commit hashes in picked snapshot builds. Multiple builds can have the same commit hash for the agent which leads to file system collisions.

Fixes https://github.com/elastic/elastic-agent/issues/4089
Fixes https://github.com/elastic/elastic-agent/issues/4091

### CODE REVIEW

**Comment by rdner** [comment]
The linter is failing in the unchanged code.

File: testing/integration/upgrade_downgrade_test.go [hunk 1]
```diff
@@ -41,25 +42,29 @@ func TestStandaloneDowngradeToSpecificSnapshotBuild(t *testing.T) {
...
+	require.NoError(t, err)
+	startVersion, err := startFixture.ExecVersion(ctx)
 	require.NoError(t, err)
-	if len(builds.Builds) < 2 {
-		t.Skipf("not enough SNAPSHOT builds exist for version %s. Found %d", latestSnapshotVersion.VersionWithPrerelease(), len(builds.Builds))
+
+	// when picking the build we need to find one that has a different commit hash to avoid file system
+	// collisions during the upgrade process.
+	// multiple builds can have different IDs but the same commit hash.
+	build, err := aac.GetLastBuildForVersion(ctx, latestSnapshotVersion.VersionWithPrerelease(), startVersion.Binary.Commit)
```
Comment (pchila) [inline testing/integration/upgrade_downgrade_test.go:59]: To properly test the specific feature we should choose a build that is not the latest SNAPSHOT build.
If we pick the latest snapshot the result after the upgrade in the form `elastic-agent upgrade x.y.z-SNAPSHOT-<buildid>` would be indistinguishable from a regular `elastic-agent upgrade x.y.z-SNAPSHOT`

**Review by pchila — CHANGES_REQUESTED** [review]
Please restore the test behavior for `TestStandaloneDowngradeToSpecificSnapshotBuild` to NOT use the latest SNAPSHOT build

File: testing/integration/upgrade_downgrade_test.go (hunk 1 above)
Comment (rdner) [reply testing/integration/upgrade_downgrade_test.go:59]: @pchila can you elaborate more? 

* We get a list of snapshot builds of the same version as we're running
* Then we look for the latest snapshot that **does not match** the current hash of `main`
* This tests upgrading to the previous SNAPSHOT

Under what circumstances there will be a top/last snapshot with a different commit hash? My idea was that the commit hash filter will always find the previous snapshot that was not built with the same commit hash, is not what we want here?

Comment (rdner) [reply testing/integration/upgrade_downgrade_test.go]: Changed in https://github.com/elastic/elastic-agent/pull/4095/commits/dd512b44169c539789719942ed60aa9fb3924c9f

File: testing/upgradetest/upgrader.go [hunk 2]
```diff
@@ -233,7 +237,7 @@ func PerformUpgrade(
 		}
 	}
 
-	logger.Logf("Upgrading from version %q (%s) to version %q (%s)", startParsedVersion, startVersionInfo.Binary.Commit, endVersionInfo.Binary.String(), endVersionInfo.Binary.Commit)
+	logger.Logf("Upgrading from version \"%s-%s\" to version \"%s-%s\"", startParsedVersion, startVersionInfo.Binary.Commit, endVersionInfo.Binary.String(), endVersionInfo.Binary.Commit)
```
Comment (rdner) [inline testing/upgradetest/upgrader.go:240]: To make it consistent with https://github.com/elastic/elastic-agent/blob/17f048008fcea105d0b86cc90743c7bfb8ce2a41/testing/integration/upgrade_fleet_test.go#L229

File: testing/integration/upgrade_downgrade_test.go [hunk 3]
```diff
@@ -53,18 +56,34 @@ func TestStandaloneDowngradeToSpecificSnapshotBuild(t *testing.T) {
...
 
-	// when picking the build we need to find one that has a different commit hash to avoid file system
-	// collisions during the upgrade process.
-	// multiple builds can have different IDs but the same commit hash.
-	build, err := aac.GetLastBuildForVersion(ctx, latestSnapshotVersion.VersionWithPrerelease(), startVersion.Binary.Commit)
-	if errors.Is(err, tools.ErrNoMatchingBuild) {
+	// We need to find a build which is not the latest (so, we can make sure we address it by a build ID
+	// in the version prefix, e.g. x.y.z-SNAPSHOT-<buildid>) and does not have the same commit hash
+	// as the currently running binary (so, we don't have a file system collision).
+	// Multiple builds can have different IDs but the same commit hash.
```
Comment (pchila) [inline testing/integration/upgrade_downgrade_test.go:62]: :+1: :heart:

**Review by pchila — APPROVED** [review]

**Review by blakerouse — APPROVED** [review]
Looks good.

**Pull request merged**